mountpoint-s3/tests/fuse_tests/cache_test.rs

```diff
@@ -0,0 +1,129 @@
+use rand_chacha::ChaChaRng;
+use std::fs;
+use std::sync::Arc;
+use std::time::Duration;
+use test_case::test_case;
+
+const CACHE_BLOCK_SIZE: u64 = 1024 * 1024;
+const CLIENT_PART_SIZE: usize = 8 * 1024 * 1024;
+
+#[test_case("key", 100, 1024; "simple")]
+#[test_case("£", 100, 1024; "non-ascii key")]
+#[test_case("key", 1024, 1024; "long key")]
+#[test_case("key", 100, 1024 * 1024; "big file")]
+fn express_cache_write_read(key_suffix: &str, key_size: usize, object_size: usize) {
+    let client = create_crt_client(CLIENT_PART_SIZE, CLIENT_PART_SIZE);
+    let bucket_name = get_standard_bucket();
+    let express_bucket_name = get_express_bucket();
+    let cache = ExpressDataCache::new(client.clone(), Default::default(), &bucket_name, &express_bucket_name);
+
+    cache_write_read_base(
+        client,
+        key_suffix,
+        key_size,
+        object_size,
+        cache,
+        "express_cache_write_read",
+    )
+}
+
+#[test_case("key", 100, 1024; "simple")]
+#[test_case("£", 100, 1024; "non-ascii key")]
+#[test_case("key", 1024, 1024; "long key")]
+#[test_case("key", 100, 1024 * 1024; "big file")]
+fn disk_cache_write_read(key_suffix: &str, key_size: usize, object_size: usize) {
+    let cache_dir = tempfile::tempdir().unwrap();
+    let cache_config = DiskDataCacheConfig {
+        block_size: CACHE_BLOCK_SIZE,
+        limit: Default::default(),
+    };
+    let cache = DiskDataCache::new(cache_dir.path().to_path_buf(), cache_config);
+
+    let client = create_crt_client(CLIENT_PART_SIZE, CLIENT_PART_SIZE);
+
+    cache_write_read_base(
+        client,
+        key_suffix,
+        key_size,
+        object_size,
+        cache,
+        "disk_cache_write_read",
+    );
+}
+
+fn cache_write_read_base<Cache>(
+    client: S3CrtClient,
+    key_suffix: &str,
+    key_size: usize,
+    object_size: usize,
+    cache: Cache,
+    test_name: &str,
+) where
+    Cache: DataCache + Send + Sync + 'static,
+{
+    let (bucket, prefix) = get_test_bucket_and_prefix(test_name);
+
+    // Mount a bucket
+    let mount_point = tempfile::tempdir().unwrap();
+    let runtime = client.event_loop_group();
+    let cache = CacheTestWrapper::new(Arc::new(cache));
+    let prefetcher = caching_prefetch(cache.clone(), runtime, Default::default());
+    let _session = create_fuse_session(
+        client,
+        prefetcher,
+        &bucket,
+        &prefix,
+        mount_point.path(),
+        Default::default(),
+    );
+
+    // Write an object, no caching happens yet
+    let key = get_object_key(&prefix, key_suffix, key_size);
+    let path = mount_point.path().join(&key);
+    let written = random_binary_data(object_size);
+    fs::write(&path, &written).expect("write should succeed");
+
+    // First read should be from the source bucket and be cached
+    let read = fs::read(&path).expect("read should succeed");
+    assert_eq!(read, written);
+
+    // Cache population is async, wait for it to happen
```

Comment:
See comment above: you are not waiting for the cache to be populated, just for `put_block` to return.

Reply:
After the `put_block` returned we [know](https://github.com/awslabs/mountpoint-s3/blob/main/mountpoint-s3/src/data_cache/express_data_cache.rs#L160) that the data is in the cache. 

Are you suggesting to wait for the `put_block` counter to change the value? e.g like this:
```rust

    pub fn wait_for_put(&self, max_wait_duration: Duration, previous_failed: u64, previous_ok: u64) {
        let st = std::time::Instant::now();
        loop {
            if st.elapsed() > max_wait_duration {
                panic!("timeout on waiting for a write to the cache to happen")
            }
            if self.put_block_failed_count.load(Ordering::SeqCst) > previous_failed
                || self.put_block_ok_count.load(Ordering::SeqCst) > previous_ok
            {
                break;
            }
            std::thread::sleep(Duration::from_millis(100));
        }
    }
```

Reply:
But yes, you should compare with the previous count in some way.

Reply:
If we are not actually using the failure counter, you could just simplify and just count all puts. We can always add counters later.

Reply:
kept 2 counters only, made it to compare with the previous valie